rr_graph_uxsdcxx.cpp

```diff
@@ -0,0 +1,3174 @@
+const char *atok_lookup_t_segment[] = {"id", "name"};
+enum class gtok_t_pin_class {PIN};
+const char *gtok_lookup_t_pin_class[] = {"pin"};
+enum class atok_t_pin_class {TYPE};
+const char *atok_lookup_t_pin_class[] = {"type"};
+enum class gtok_t_metadata {META};
+const char *gtok_lookup_t_metadata[] = {"meta"};
+enum class gtok_t_channels {CHANNEL, X_LIST, Y_LIST};
+const char *gtok_lookup_t_channels[] = {"channel", "x_list", "y_list"};
+enum class gtok_t_grid_locs {GRID_LOC};
+const char *gtok_lookup_t_grid_locs[] = {"grid_loc"};
+enum class gtok_t_connection_boxes {CONNECTION_BOX};
+const char *gtok_lookup_t_connection_boxes[] = {"connection_box"};
+enum class atok_t_connection_boxes {X_DIM, Y_DIM, NUM_BOXES};
+const char *atok_lookup_t_connection_boxes[] = {"x_dim", "y_dim", "num_boxes"};
+enum class gtok_t_block_type {PIN_CLASS};
+const char *gtok_lookup_t_block_type[] = {"pin_class"};
+enum class atok_t_block_type {ID, NAME, WIDTH, HEIGHT};
+const char *atok_lookup_t_block_type[] = {"id", "name", "width", "height"};
+enum class gtok_t_node {LOC, TIMING, SEGMENT, METADATA};
+const char *gtok_lookup_t_node[] = {"loc", "timing", "segment", "metadata"};
+enum class atok_t_node {ID, TYPE, DIRECTION, CAPACITY};
+const char *atok_lookup_t_node[] = {"id", "type", "direction", "capacity"};
+enum class gtok_t_edge {METADATA};
+const char *gtok_lookup_t_edge[] = {"metadata"};
+enum class atok_t_edge {ID, SRC_NODE, SINK_NODE, SWITCH_ID};
+const char *atok_lookup_t_edge[] = {"id", "src_node", "sink_node", "switch_id"};
+enum class gtok_t_switches {SWITCH};
+const char *gtok_lookup_t_switches[] = {"switch"};
+enum class gtok_t_segments {SEGMENT};
+const char *gtok_lookup_t_segments[] = {"segment"};
+enum class gtok_t_block_types {BLOCK_TYPE};
+const char *gtok_lookup_t_block_types[] = {"block_type"};
+enum class gtok_t_rr_nodes {NODE};
+const char *gtok_lookup_t_rr_nodes[] = {"node"};
+enum class gtok_t_rr_edges {EDGE};
+const char *gtok_lookup_t_rr_edges[] = {"edge"};
+enum class gtok_t_rr_graph {BINARY_NODES_AND_EDGES, CONNECTION_BOXES, CHANNELS, SWITCHES, SEGMENTS, BLOCK_TYPES, GRID, RR_NODES, RR_EDGES};
+const char *gtok_lookup_t_rr_graph[] = {"binary_nodes_and_edges", "connection_boxes", "channels", "switches", "segments", "block_types", "grid", "rr_nodes", "rr_edges"};
+enum class atok_t_rr_graph {TOOL_NAME, TOOL_VERSION, TOOL_COMMENT};
+const char *atok_lookup_t_rr_graph[] = {"tool_name", "tool_version", "tool_comment"};
+
+
+/* Lexing functions. These convert the const char *s of PugiXML to enums.
+ * You may find numerous "break"s there. Without them, a warning explosion ensues. */
+inline enum_switch_type lex_switch_type(const char *in, bool throw_on_invalid){
+	unsigned int len = strlen(in);
+	switch(len){
+	case 3:
+		switch(in[0]){
+		case onechar('m', 0, 8):
+			switch(in[1]){
+			case onechar('u', 0, 8):
+				switch(in[2]){
+				case onechar('x', 0, 8):
+					return enum_switch_type::MUX;
+				break;
+				default: break;
```

Comment:
The formatting in this is pretty weird....

Reply:
I think I hacked in the `default:break` to stop GCC from complaining. https://github.com/duck2/uxsdcxx/blob/master/third_party/triehash.py#L157